File: Merge_sort.kt
fun mergeSort(arr: IntArray, left: Int, right: Int)
{
    // 배열을 가장 작은 단위까지 나눔
    if (left < right) {
        val middle = left + (right - left) / 2

        // 왼쪽과 오른쪽 부분 배열 정렬
        mergeSort(arr, left, middle)
        mergeSort(arr, middle + 1, right)

        // 병합
        merge(arr, left, middle, right)
    }
}

fun merge(arr: IntArray, left: Int, middle: Int, right: Int) {
    val n1 = middle - left + 1
    val n2 = right - middle

    // 임시 배열 생성
    val L = IntArray(n1)
    val R = IntArray(n2)

    // 데이터를 임시 배열로 복사
    for (i in 0 until n1) {
        L[i] = arr[left + i]
    }
    for (j in 0 until n2) {
        R[j] = arr[middle + 1 + j]
    }

    // 임시 배열들을 병합하여 원래 배열에 복사
    var i = 0
    var j = 0
    var k = left
    // 왼쪽 배열과 오른쪽 배열중 작은 값을 결과 배열에 저장
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i]
            i++
        } else {
            arr[k] = R[j]
            j++
        }
        k++
    }
    // 나머지 요소들을 원래 배열에 복사
    while (i < n1) {
        arr[k] = L[i]
        i++
        k++
    }
    while (j < n2) {
        arr[k] = R[j]
        j++
        k++
    }
}

fun main() {
    print("Enter numbers separated by spaces: ")
    val input = readLine() ?: return
    val arr = input.split(" ").map { it.toIntOrNull() ?: return }.toIntArray()

    println("Original array: ${arr.contentToString()}")
    mergeSort(arr, 0, arr.size - 1)
    println("Sorted array: ${arr.contentToString()}")
}

File: merge_sort.cpp
#include <iostream>
#include <stack>
#include <vector>
#include <queue>
#include <algorithm>
#include <string>

int temp[1000000];
int* arr;
using namespace std;
void merge(int left, int mid, int right)
{
    int l, m, t;

    l = left;
    m = mid + 1;
    t = left;

    //정렬 후 합병
    while (l <= mid && m <= right)
    {
        if (arr[l] <= arr[m])
            temp[t++] = arr[l++];
        else
            temp[t++] = arr[m++];
    }

    if (l > mid)
    {
        while (m <= right)
            temp[t++] = arr[m++];
    }
    else
    {
        while (l <= mid)
            temp[t++] = arr[l++];
    }

    //원래 배열로 복사
    while (left <= right)
    {
        arr[left] = temp[left];
        left++;
    }
}

void merge_sort(int left, int right)// 분할과정 -> partition
{
    int mid;

    if (left >= right)
        return;

    mid = (left + right) / 2;
    merge_sort(left, mid);
    merge_sort(mid + 1, right);
    merge(left, mid, right);
}

int main() {
    int n;

    cout << "배열의 크기 입력: ";
    cin >> n;

    arr = new int[n];//동적메모리 할당

    cout << "배열의 요소 입력: ";
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    merge_sort(0, n - 1);

    cout << "정렬된 배열: ";
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;

    delete[] arr; // 동적 할당 메모리 해제

    return 0;
}